cache/memoize.ts

```diff
@@ -0,0 +1,227 @@
+    _serializeArgList(
+      cache as MemoizationCache<unknown, unknown>,
+    ) as unknown as (
+      (this: ThisParameterType<Fn>, ...args: Parameters<Fn>) => Key
+    );
+  const truncateArgs = options?.truncateArgs ?? false;
+  const cacheRejectedPromises = options?.cacheRejectedPromises ?? false;
+
+  const memoized = function (
+    this: ThisParameterType<Fn>,
+    ...args: Parameters<Fn>
+  ): ReturnType<Fn> {
+    if (truncateArgs) args = args.slice(0, fn.length) as Parameters<Fn>;
+
+    const key = getKey.apply(this, args) as Key;
+
+    if (cache.has(key)) {
+      return cache.get(key)!;
+    }
+
+    let val = fn.apply(this, args) as ReturnType<Fn>;
+
+    if (val instanceof Promise && !cacheRejectedPromises) {
+      val = val.catch((reason) => {
+        cache.delete(key);
+        throw reason;
+      }) as typeof val;
+    }
+
+    cache.set(key, val);
+
+    return val;
+  } as Fn;
+
+  return Object.defineProperties(
+    Object.assign(memoized, { cache: cache as Cache, getKey }),
+    {
+      length: { value: fn.length },
+      name: { value: fn.name },
+    },
+  );
+}
+
+/**
+ * Default serialization of arguments list for use as cache keys. Equivalence
+ * follows [`SameValueZero`](https://tc39.es/ecma262/multipage/abstract-operations.html#sec-samevaluezero)
+ * reference equality, such that `getKey(x, y) === getKey(x, y)` for all values
+ * of `x` and `y`, but `getKey({}) !== getKey({})`.
+ *
+ * @param cache - The cache for which the keys will be used.
+ * @returns `getKey`, the function for getting cache keys.
+ */
+export function _serializeArgList<Return>(
+  cache: MemoizationCache<unknown, Return>,
+): (this: unknown, ...args: unknown[]) => string {
+  const weakKeyToKeySegmentCache = new WeakMap<WeakKey, string>();
+  const weakKeySegmentToKeyCache = new Map<string, string[]>();
+  let i = 0;
+
+  const registry = new FinalizationRegistry<string>((keySegment) => {
+    for (const key of weakKeySegmentToKeyCache.get(keySegment) ?? []) {
+      cache.delete(key);
+    }
+    weakKeySegmentToKeyCache.delete(keySegment);
+  });
+
+  return function getKey(...args) {
+    const weakKeySegments: string[] = [];
+    const keySegments = [this, ...args].map((arg) => {
+      if (typeof arg === "undefined") return "undefined";
+      if (typeof arg === "bigint") return `${arg}n`;
+
+      if (typeof arg === "number") {
+        return String(arg);
+      }
+
+      if (
+        arg === null ||
+        typeof arg === "string" ||
+        typeof arg === "boolean"
+      ) {
+        // This branch will need to be updated if further types are added to
+        // the language that support value equality,
+        // e.g. https://github.com/tc39/proposal-record-tuple
+        return JSON.stringify(arg);
+      }
+
+      try {
+        assertWeakKey(arg);
+      } catch (e) {
+        if (typeof arg === "symbol") {
+          return `Symbol.for(${JSON.stringify(arg.description)})`;
+        }
+        throw e;
```

Comment:
Non-blocking: Can you please add a test for this line?

Reply:
Replacing the re-throw with `unreachable` as this line should never be hit.